File: backend/app/api/v1/endpoints/alertas.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from uuid import UUID

from app.core.deps import get_db, get_usuario_actual, require_admin_or_supervisor
from app.models.usuario import Usuario
from app.models.alerta import TipoAlerta, PrioridadAlerta
from app.schemas.alerta import (
    AlertaResponse, AlertaConteo, VerificacionResultado
)
from app.services import alertas_service

router = APIRouter()


@router.get("", response_model=List[AlertaResponse])
def listar_alertas(
    solo_no_leidas: bool = False,
    solo_no_resueltas: bool = True,
    tipo: Optional[str] = None,
    prioridad: Optional[str] = None,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(get_usuario_actual)
):
    """Lista las alertas del usuario."""
    tipo_enum = TipoAlerta(tipo) if tipo else None
    prioridad_enum = PrioridadAlerta(prioridad) if prioridad else None

    alertas = alertas_service.get_alertas(
        db,
        usuario_id=usuario.id,
        solo_no_leidas=solo_no_leidas,
        solo_no_resueltas=solo_no_resueltas,
        tipo=tipo_enum,
        prioridad=prioridad_enum,
        skip=skip,
        limit=limit
    )

    return [
        AlertaResponse(
            id=a.id,
            tipo=a.tipo,
            prioridad=a.prioridad,
            titulo=a.titulo,
            mensaje=a.mensaje,
            referencia_tipo=a.referencia_tipo,
            referencia_id=a.referencia_id,
            usuario_id=a.usuario_id,
            leida=a.leida,
            resuelta=a.resuelta,
            fecha_resolucion=a.fecha_resolucion,
            resuelto_por_id=a.resuelto_por_id,
            created_at=a.created_at
        )
        for a in alertas
    ]


@router.get("/conteo", response_model=AlertaConteo)
def contar_alertas(
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(get_usuario_actual)
):
    """Obtiene el conteo de alertas por prioridad."""
    alertas = alertas_service.get_alertas(
        db,
        usuario_id=usuario.id,
        solo_no_resueltas=True,
        limit=500
    )

    return AlertaConteo(
        total=len(alertas),
        criticas=sum(1 for a in alertas if a.prioridad == 'critica'),
        altas=sum(1 for a in alertas if a.prioridad == 'alta'),
        medias=sum(1 for a in alertas if a.prioridad == 'media'),
        bajas=sum(1 for a in alertas if a.prioridad == 'baja'),
        no_leidas=sum(1 for a in alertas if not a.leida)
    )


@router.post("/{alerta_id}/leer", response_model=AlertaResponse)
def marcar_leida(
    alerta_id: UUID,
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(get_usuario_actual)
):
    """Marca una alerta como leída."""
    alerta = alertas_service.marcar_como_leida(db, alerta_id)
    if not alerta:
        raise HTTPException(status_code=404, detail="Alerta no encontrada")

    return AlertaResponse(
        id=alerta.id,
        tipo=alerta.tipo,
        prioridad=alerta.prioridad,
        titulo=alerta.titulo,
        mensaje=alerta.mensaje,
        referencia_tipo=alerta.referencia_tipo,
        referencia_id=alerta.referencia_id,
        usuario_id=alerta.usuario_id,
        leida=alerta.leida,
        resuelta=alerta.resuelta,
        fecha_resolucion=alerta.fecha_resolucion,
        resuelto_por_id=alerta.resuelto_por_id,
        created_at=alerta.created_at
    )


@router.post("/leer-todas")
def marcar_todas_leidas(
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(get_usuario_actual)
):
    """Marca todas las alertas como leídas."""
    cantidad = alertas_service.marcar_todas_leidas(db, usuario.id)
    return {"message": f"Se marcaron {cantidad} alertas como leídas"}


@router.post("/{alerta_id}/resolver", response_model=AlertaResponse)
def resolver_alerta(
    alerta_id: UUID,
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(get_usuario_actual)
):
    """Marca una alerta como resuelta."""
    alerta = alertas_service.resolver_alerta(db, alerta_id, usuario.id)
    if not alerta:
        raise HTTPException(status_code=404, detail="Alerta no encontrada")

    return AlertaResponse(
        id=alerta.id,
        tipo=alerta.tipo,
        prioridad=alerta.prioridad,
        titulo=alerta.titulo,
        mensaje=alerta.mensaje,
        referencia_tipo=alerta.referencia_tipo,
        referencia_id=alerta.referencia_id,
        usuario_id=alerta.usuario_id,
        leida=alerta.leida,
        resuelta=alerta.resuelta,
        fecha_resolucion=alerta.fecha_resolucion,
        resuelto_por_id=alerta.resuelto_por_id,
        created_at=alerta.created_at
    )


@router.post("/verificar", response_model=VerificacionResultado)
def ejecutar_verificaciones(
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(require_admin_or_supervisor)
):
    """Ejecuta todas las verificaciones y genera alertas (solo admin/supervisor)."""
    resultado = alertas_service.ejecutar_todas_verificaciones(db)
    return VerificacionResultado(**resultado)


@router.get("/tipos")
def obtener_tipos():
    """Obtiene los tipos de alerta disponibles."""
    return [
        {"value": t.value, "label": t.value.replace("_", " ").title()}
        for t in TipoAlerta
    ]


@router.get("/prioridades")
def obtener_prioridades():
    """Obtiene las prioridades disponibles."""
    return [
        {"value": p.value, "label": p.value.title()}
        for p in PrioridadAlerta
    ]

File: frontend/src/services/alertas.ts
import api from './api'

export type TipoAlerta =
  | 'stock_minimo'
  | 'demora_etapa'
  | 'etapa_pendiente'
  | 'limite_credito'
  | 'equipo_mantenimiento'
  | 'proyecto_vencido'
  | 'pago_pendiente'

export type PrioridadAlerta = 'baja' | 'media' | 'alta' | 'critica'

export interface Alerta {
  id: string
  tipo: TipoAlerta
  prioridad: PrioridadAlerta
  titulo: string
  mensaje: string
  referencia_tipo?: string
  referencia_id?: string
  usuario_id?: string
  leida: boolean
  resuelta: boolean
  fecha_resolucion?: string
  resuelto_por_id?: string
  created_at: string
}

export interface AlertaConteo {
  total: number
  criticas: number
  altas: number
  medias: number
  bajas: number
  no_leidas: number
}

export interface VerificacionResultado {
  stock_minimo: number
  demoras_etapas: number
  equipos_mantenimiento: number
  limites_credito: number
  proyectos_vencidos: number
  total: number
}

export const alertasService = {
  getAlertas: async (params?: {
    solo_no_leidas?: boolean
    solo_no_resueltas?: boolean
    tipo?: TipoAlerta
    prioridad?: PrioridadAlerta
    skip?: number
    limit?: number
  }): Promise<Alerta[]> => {
    const response = await api.get('/alertas', { params })
    return response.data
  },

  getConteo: async (): Promise<AlertaConteo> => {
    const response = await api.get('/alertas/conteo')
    return response.data
  },

  marcarLeida: async (alertaId: string): Promise<Alerta> => {
    const response = await api.post(`/alertas/${alertaId}/leer`)
    return response.data
  },

  marcarTodasLeidas: async (): Promise<{ message: string }> => {
    const response = await api.post('/alertas/leer-todas')
    return response.data
  },

  resolverAlerta: async (alertaId: string): Promise<Alerta> => {
    const response = await api.post(`/alertas/${alertaId}/resolver`)
    return response.data
  },

  ejecutarVerificaciones: async (): Promise<VerificacionResultado> => {
    const response = await api.post('/alertas/verificar')
    return response.data
  },

  getTipos: async (): Promise<{ value: string; label: string }[]> => {
    const response = await api.get('/alertas/tipos')
    return response.data
  },

  getPrioridades: async (): Promise<{ value: string; label: string }[]> => {
    const response = await api.get('/alertas/prioridades')
    return response.data
  }
}

// Helpers
export const getPrioridadColor = (prioridad: PrioridadAlerta): string => {
  const colors = {
    baja: 'text-gray-500 bg-gray-100',
    media: 'text-blue-500 bg-blue-100',
    alta: 'text-orange-500 bg-orange-100',
    critica: 'text-red-500 bg-red-100'
  }
  return colors[prioridad] || colors.media
}

export const getTipoIcon = (tipo: TipoAlerta): string => {
  const icons = {
    stock_minimo: 'Package',
    demora_etapa: 'Clock',
    etapa_pendiente: 'AlertTriangle',
    limite_credito: 'CreditCard',
    equipo_mantenimiento: 'Wrench',
    proyecto_vencido: 'Calendar',
    pago_pendiente: 'DollarSign'
  }
  return icons[tipo] || 'Bell'
}

export const getTipoLabel = (tipo: TipoAlerta): string => {
  const labels = {
    stock_minimo: 'Stock Mínimo',
    demora_etapa: 'Demora en Etapa',
    etapa_pendiente: 'Etapa Pendiente',
    limite_credito: 'Límite de Crédito',
    equipo_mantenimiento: 'Mantenimiento Equipo',
    proyecto_vencido: 'Proyecto Vencido',
    pago_pendiente: 'Pago Pendiente'
  }
  return labels[tipo] || tipo
}
